libcodechecker/log/log_parser.py

```diff
@@ -0,0 +1,791 @@
+    def item(self):
+        return self._item
+
+
+def get_language(extension):
+    # TODO: There are even more in the man page of gcc.
+    mapping = {'.c': 'c',
+               '.cp': 'c++',
+               '.cpp': 'c++',
+               '.cxx': 'c++',
+               '.txx': 'c++',
+               '.cc': 'c++',
+               '.C': 'c++',
+               '.ii': 'c++',
+               '.m': 'objective-c',
+               '.mm': 'objective-c++'}
+    return mapping.get(extension)
+
+
+def merge_space_files(sources, filename):
+    sub = filename.split()
+    ls = len(sub)
+
+    if ls == 0:
+        return
+
+    i = 0
+    while i < len(sources) - ls + 1:
+        if sources[i:i + ls] == sub:
+            del sources[i:i + ls]
+            sources.insert(i, filename)
+        i += 1
+
+
+def __append_from_file(flag_iterator, buildaction):
+    if flag_iterator.item == '-filelist':
+        next(flag_iterator)
+
+        with open(flag_iterator.item) as f:
+            for line in f:
+                buildaction.sources.append(line.strip())
+
+        return True
+
+    return False
+
+
+def __collect_compile_opts(flag_iterator, buildaction):
+    """
+    This function collects the compilation (i.e. not linker or preprocessor)
+    flags to the buildaction.
+    """
+    if COMPILE_OPTIONS.match(flag_iterator.item):
+        buildaction.analyzer_options.append(flag_iterator.item)
+        return True
+
+    m = COMPILE_OPTIONS_MERGED.match(flag_iterator.item)
+
+    if m:
+        flag = m.group(0)
+        together = len(flag) != len(flag_iterator.item)
+
+        if together:
+            param = flag_iterator.item[len(flag):]
+        else:
+            next(flag_iterator)
+            param = flag_iterator.item
+
+        if flag == '-I':
+            param = os.path.normpath(
+                os.path.join(buildaction.directory, param))
+
+        if together:
+            buildaction.analyzer_options.append(flag + param)
+        else:
+            buildaction.analyzer_options.extend([flag, param])
+
+        return True
+
+    return False
+
+
+def __collect_sources(flag_iterator, buildaction):
+    """
+    This function collects the compiled source file names (i.e. the arguments
+    which don't start with a dash character) to the buildaction object.
+    """
+    if flag_iterator.item[0] != '-':
+        buildaction.sources.append(flag_iterator.item.strip('"'))
+        return True
+
+    return False
+
+
+def __determine_action_type(flag_iterator, buildaction):
+    """
+    This function determines whether this is a preprocessing, compilation or
+    linking action and sets it in the buildaction object.
+    """
+    if flag_iterator.item == '-c':
+        buildaction.action_type = BuildAction.COMPILE
+        return True
+    elif flag_iterator.item.startswith('-print-prog-name'):
+        buildaction.action_type = BuildAction.INFO
+        return True
+    elif re.match('-(E|M[T|Q|F|J|P|V|M]*)$', flag_iterator.item):
```

Comment:
This is also hot code. I think we should compile this regular expression once when the module is loaded and use the compiled object to do the matching here.